File: 0063-unique-paths-ii/0063-unique-paths-ii.py
class Solution:

    def dfs(self,obstacleGrid,i,j,dp): 
        m,n = len(obstacleGrid),len(obstacleGrid[0])
        if i<0 or j<0 or i==m or j==n or obstacleGrid[i][j]==1:
            return 0
        if i==m-1 and j==n-1:
            return 1 
        if dp[i][j]!=-1: 
            return dp[i][j]

        dp[i][j]=self.dfs(obstacleGrid,i+1,j,dp) +  self.dfs(obstacleGrid,i,j+1,dp)
        return dp[i][j]

    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        # i,j -> bottom or right 
        # no reursive loops & dp to keep track of repeated positions 
        i,j = 0,0 
        m,n = len(obstacleGrid),len(obstacleGrid[0])
        dp = [ [-1 for i in range(n)] for _ in range(m)]
        return self.dfs(obstacleGrid,i,j,dp)

File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& grid) {
        vector<vector<int>> dp(grid.size(),vector<int>(grid[0].size(),-1));
        return dfs(grid,0,0,dp);
    }

    int dfs(vector<vector<int>>& grid,int x,int y,vector<vector<int>>& dp){
        if(x==grid.size() || y==grid[0].size() || grid[x][y]==1) return 0;
        if(x==grid.size()-1 && y==grid[0].size()-1) return 1;
        if(dp[x][y]!=-1) return dp[x][y]; 
        // if()
        return dp[x][y]=dfs(grid,x+1,y,dp)+dfs(grid,x,y+1,dp);
    }

};